### PR TITLE
ap-4712 # Added crop functionality to camera element

### PR DESCRIPTION
**Pull request by Zaxist**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

### CODE REVIEW

File: src/components/renderer/AnnotationModal.tsx [hunk 1]
```diff
@@ -169,3 +170,36 @@ function AnnotationModal({
 }
 
 export default React.memo(AnnotationModal)
+
+export const superimposeAnnotationOnImage = async ({
```
Comment (mymattcarroll) [inline src/components/renderer/AnnotationModal.tsx:174]: Should this be used in this file?

File: src/form-elements/FormElementFile.tsx [hunk 2]
```diff
@@ -41,6 +70,69 @@ const FormElementFile = ({
...
+    [attachmentUrl, clearIsAnnotating, element, file, onChange],
+  )
+  const handleSaveCrop = React.useCallback(
+    async (croppedAreaPixels: Area) => {
+      if (!attachmentUrl || file.type) return
+      clearIsCropping()
+
+      setState({
+        isLoading: true,
+      })
```
Comment (mymattcarroll) [inline src/form-elements/FormElementFile.tsx:109]: Same here

File: src/form-elements/FormElementFile.tsx [hunk 3]
```diff
@@ -56,22 +148,68 @@ const FormElementFile = ({
...
+        attachmentUrl={attachmentUrl}
+        isContentTypeImage={isContentTypeImage}
+        fileName={file.fileName}
+        onDownload={canDownload ? handleDownload : undefined}
+        onRemove={handleRemove}
+        onRetry={handleRetry}
+        progress={progress}
+        index={index}
+        onAnnotate={setIsAnnotating}
+        onCrop={setIsCropping}
```
[the remark this answers is not in the document]
Comment (Zaxist) [reply src/form-elements/FormElementFile.tsx:168]: Yep, meant to come back to that.

File: src/form-elements/FormElementFile.tsx [hunk 4]
```diff
@@ -41,6 +70,69 @@ const FormElementFile = ({
...
   }, [file])
 
+  const handleAnnotate = React.useCallback(
+    async (annotationDataUri: string) => {
+      if (!attachmentUrl || file.type) return
+      clearIsAnnotating()
+
+      setState({
+        isLoading: true,
+      })
```
[the remark this answers is not in the document]
Comment (Zaxist) [reply src/form-elements/FormElementFile.tsx:80]: Meh, I don't think the more verbose state setting code is worth it.

File: src/components/ImageCropper/CropModal.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+function CropModal({
+  imageSrc,
+  onClose,
+  onSave,
+}: {
+  imageSrc: string
+  onClose: () => void
+  onSave: (imageArea: Area) => void
+}) {
+  const annotationContentElementRef = React.useRef<HTMLDivElement>(null)
```
Comment (mymattcarroll) [inline src/components/ImageCropper/CropModal.tsx:15]: Is this used?

**Review by mymattcarroll — APPROVED** [review]
